x/distribution/keeper/keeper.go

```diff
@@ -35,123 +35,53 @@ func NewKeeper(cdc *codec.Codec, key sdk.StoreKey, paramSpace params.Subspace, c
-	store := ctx.KVStore(k.storeKey)
-	b := store.Get(FeePoolKey)
-	if b == nil {
-		panic("Stored fee pool should not have been nil")
+// withdraw rewards from a delegation
+func (k Keeper) WithdrawDelegationRewards(ctx sdk.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) sdk.Error {
+	val := k.stakingKeeper.Validator(ctx, valAddr)
+	if val == nil {
+		return types.ErrNoValidatorDistInfo(k.codespace)
 	}
-	k.cdc.MustUnmarshalBinaryLengthPrefixed(b, &feePool)
-	return
-}
-
-// set the global fee pool distribution info
-func (k Keeper) SetFeePool(ctx sdk.Context, feePool types.FeePool) {
-	store := ctx.KVStore(k.storeKey)
-	b := k.cdc.MustMarshalBinaryLengthPrefixed(feePool)
-	store.Set(FeePoolKey, b)
-}
-
-// get the total validator accum for the ctx height
-// in the fee pool
-func (k Keeper) GetFeePoolValAccum(ctx sdk.Context) sdk.Dec {
 
-	// withdraw self-delegation
-	height := ctx.BlockHeight()
-	totalPower := sdk.NewDecFromInt(k.stakingKeeper.GetLastTotalPower(ctx))
-	fp := k.GetFeePool(ctx)
-	return fp.GetTotalValAccum(height, totalPower)
-}
-
-//______________________________________________________________________
-
-// set the proposer public key for this block
-func (k Keeper) GetPreviousProposerConsAddr(ctx sdk.Context) (consAddr sdk.ConsAddress) {
-	store := ctx.KVStore(k.storeKey)
-
-	b := store.Get(ProposerKey)
-	if b == nil {
-		panic("Previous proposer not set")
+	del := k.stakingKeeper.Delegation(ctx, delAddr, valAddr)
+	if del == nil {
+		return types.ErrNoDelegationDistInfo(k.codespace)
 	}
 
-	k.cdc.MustUnmarshalBinaryLengthPrefixed(b, &consAddr)
-	return
-}
-
-// get the proposer public key for this block
-func (k Keeper) SetPreviousProposerConsAddr(ctx sdk.Context, consAddr sdk.ConsAddress) {
-	store := ctx.KVStore(k.storeKey)
-	b := k.cdc.MustMarshalBinaryLengthPrefixed(consAddr)
-	store.Set(ProposerKey, b)
-}
-
-//______________________________________________________________________
-
-// get context required for withdraw operations
-func (k Keeper) GetWithdrawContext(ctx sdk.Context,
-	valOperatorAddr sdk.ValAddress) types.WithdrawContext {
+	// withdraw rewards
+	if err := k.withdrawDelegationRewards(ctx, val, del); err != nil {
+		return err
+	}
 
-	feePool := k.GetFeePool(ctx)
-	height := ctx.BlockHeight()
-	validator := k.stakingKeeper.Validator(ctx, valOperatorAddr)
-	lastValPower := k.stakingKeeper.GetLastValidatorPower(ctx, valOperatorAddr)
-	lastTotalPower := sdk.NewDecFromInt(k.stakingKeeper.GetLastTotalPower(ctx))
+	// reinitialize the delegation
+	k.initializeDelegation(ctx, valAddr, delAddr)
 
-	return types.NewWithdrawContext(
-		feePool, height, lastTotalPower, sdk.NewDecFromInt(lastValPower),
-		validator.GetCommission())
+	return nil
 }
 
-//______________________________________________________________________
-// PARAM STORE
+// withdraw validator commission
+func (k Keeper) WithdrawValidatorCommission(ctx sdk.Context, valAddr sdk.ValAddress) sdk.Error {
```

Comment:
Same RE:placement - I really see this function in `validator.go`

Reply:
https://github.com/cosmos/cosmos-sdk/pull/3099#discussion_r248273121